sfaira/estimators/keras.py

```diff
@@ -556,75 +559,51 @@ def _get_dataset(
         if idx is None:
             idx = np.arange(0, self.data.n_obs)
 
-        if mode == 'train' or mode == 'train_val':
+        if mode in ['train', 'train_val', 'eval', 'predict']:
             # Prepare data reading according to whether anndata is backed or not:
-            if self.data.isbacked:
-                n_features = self.data.X.shape[1]
-                output_types, output_shapes = self._get_output_dim(n_features, model_type)
-
-                if model_type == "vae":
-                    def generator():
-                        sparse = isinstance(self.data.X[0, :], scipy.sparse.spmatrix)
-                        for i in idx:
-                            # (_,_), (_,sf) is dummy for kl loss
-                            x = self.data.X[i, :].toarray().flatten() if sparse else self.data.X[i, :].flatten()
-                            sf = self._prepare_sf(x=x)[0]
-                            yield (x, sf), (x, sf)
-                else:
-                    def generator():
-                        sparse = isinstance(self.data.X[0, :], scipy.sparse.spmatrix)
-                        for i in idx:
-                            x = self.data.X[i, :].toarray().flatten() if sparse else self.data.X[i, :].flatten()
-                            sf = self._prepare_sf(x=x)[0]
-                            yield (x, sf), x
-            else:
-                x = self._prepare_data_matrix(idx=idx)
-                sf = self._prepare_sf(x=x)
-                n_features = x.shape[1]
-                output_types, output_shapes = self._get_output_dim(n_features, model_type)
-
-                if model_type == "vae":
-                    def generator():
-                        for i in range(x.shape[0]):
-                            # (_,_), (_,sf) is dummy for kl loss
-                            yield (x[i, :].toarray().flatten(), sf[i]), (x[i, :].toarray().flatten(), sf[i])
-                else:
-                    def generator():
-                        for i in range(x.shape[0]):
-                            yield (x[i, :].toarray().flatten(), sf[i]), x[i, :].toarray().flatten()
+            x = self.data.X if self.data.isbacked else self._prepare_data_matrix(idx=idx)
+
+            def convert_sparse_matrix_to_sparse_tensor(x):
+                coo = x.tocoo()
+                indices = np.mat([coo.row, coo.col]).transpose()
+                return tf.SparseTensor(indices, coo.data, coo.shape)
+
+            def generator():
+                is_sparse = isinstance(x[0, :], scipy.sparse.spmatrix)
+                indices = idx if self.data.isbacked else range(x.shape[0])
+                for i in indices:
+                    x_sample = x[i, :].toarray().flatten() if is_sparse else x[i, :].flatten()
+                    if not isinstance(x_sample, np.ndarray):  # Convert to dense tensor if sparse object
```

Comment:
is this ever triggered? you use toarray above. also you convert a data to a sparse tensor and then immidiately turn that into a dense tensor, is that really useful?

Reply:
Good point, this check is from before and might not get triggered. Regarding tensor conversion: I think this is necessary as (apparently) there's no way to directly convert a sparse matrix to dense tensor.

Reply:
but is `scipy sparse -> tf sparse -> tf dense` better than `scipy sparse -> np dense -> tf dense`? I somehow doubt this because the latter implies that the generator is only numpy/scipy code which should be integrated well

Reply:
I check, the conversion is unnecessary.